docs/configuration.md
Configuration Reference
=======================

Here you will find all configuration options of `sylius_grid`.

<details open><summary>Yaml</summary>

```yaml
sylius_grid:
    grids:
        app_user: # Your grid name
            driver:
                name: doctrine/orm
                options:
                    class: "%app.model.user.class%"
                    repository:
                        method: myCustomMethod
                        arguments:
                            id: resource.id
            sorting:
                name: asc
            limits: [10, 25, 50, 100]
            fields:
                name:
                    type: twig # Type of field
                    label: Name # Label
                    path: . # dot means a whole object
                    sortable: ~ | field path
                    position: 100
                    options:
                        template: "@Grid/Column/_name.html.twig" # Only twig column
                        vars:
                            labels: # a template of how does the label look like
                    enabled: true
            filters:
                name:
                    type: string # Type of filter
                    label: app.ui.name
                    enabled: true
                    template: ~
                    position: 100
                    options:
                        fields: { }
                    form_options:
                        type: contains # type of string filtering option, if you one to have just one
                    default_value:
                        type: contains
                        value: example
                enabled:
                    type: boolean # Type of filter
                    label: app.ui.enabled
                    enabled: true
                    template: ~
                    position: 100
                    options:
                        field: enabled
                    form_options: { }
                    default_value: 'false'
                date:
                    type: date # Type of filter
                    label: app.ui.created_at
                    enabled: true
                    template: ~
                    position: 100
                    options:
                        field: createdAt
                    form_options: { }
                    default_value: ~
                channel:
                    type: entity # Type of filter
                    label: app.ui.channel
                    enabled: true
                    template: ~
                    position: 100
                    options:
                        fields: [channel]
                    form_options:
                        class: "%app.model.channel.class%"
                    default_value: ~
            actions:
                main:
                    create:
                        type: create
                        label: sylius.ui.create
                        enabled: true
                        icon: ~
                        position: 100
                item:
                    update:
                        type: update
                        label: sylius.ui.edit
                        enabled: true
                        icon: ~
                        position: 100
                        options: { }
                    delete:
                        type: delete
                        label: sylius.ui.delete
                        enabled: true
                        icon: ~
                        position: 100
                        options: { }
                    show:
                        type: show
                        label: sylius.ui.show
                        enabled: true
                        icon: ~
                        position: 100
                        options:
                            link:
                                route: app_user_show
                                parameters:
                                    id: resource.id
                    archive:
                        type: archive
                        label: sylius.ui.archive
                        enabled: true
                        icon: ~
                        position: 100
                        options:
                            restore_label: sylius.ui.restore
                bulk:
                    delete:
                        type: delete
                        label: sylius.ui.delete
                        enabled: true
                        icon: ~
                        position: 100
                        options: { }
                subitem:
                    addresses:
                        type: links
                        label: sylius.ui.manage_addresses
                        options:
                            icon: cubes
                            links:
                                index:
                                    label: sylius.ui.list_addresses
                                    icon: list
                                    route: app_admin_user_address_index
                                    visible: resource.hasAddress
                                    parameters:
                                        userId: resource.id
                                create:
                                    label: sylius.ui.generate
                                    icon: random
                                    route: app_admin_user_address_create
                                    parameters:
                                        userId: resource.id
```

</details>

<details open><summary>PHP</summary>

```php
<?php

use Sylius\Bundle\GridBundle\Builder\Action\Action;
use Sylius\Bundle\GridBundle\Builder\ActionGroup\BulkActionGroup;
use Sylius\Bundle\GridBundle\Builder\ActionGroup\ItemActionGroup;
use Sylius\Bundle\GridBundle\Builder\ActionGroup\MainActionGroup;
use Sylius\Bundle\GridBundle\Builder\ActionGroup\SubItemActionGroup;
use Sylius\Bundle\GridBundle\Builder\Field\Field;
use Sylius\Bundle\GridBundle\Builder\Filter\Filter;
use Sylius\Bundle\GridBundle\Builder\GridBuilder;
use Sylius\Bundle\GridBundle\Config\GridConfig;

return static function (GridConfig $grid): void {
    $grid->addGrid(GridBuilder::create('app_user', '%app.model.user.class%')
        ->setDriver('doctrine/orm')
        ->setRepositoryMethod('myCustomMethod', ['id' => 'resource.id'])
        ->orderBy('name', 'asc')
        ->setLimits([10, 25, 50, 100])
        ->addField(
            Field::create('name', 'twig') // Name & Type of field
                ->setLabel('Name') // # Label
                ->setPath('.') // dot means a whole object
                ->setSortable(true)
                ->setPosition(100)
                ->setOptions([
                    'template' => '@Grid/Column/_name.html.twig', // Only twig column
                ])
                ->setEnabled(true)
        )
        ->addFilter(
            Filter::create('name', 'string') // Name & Type of filter
                ->setLabel('app.ui.name')
                ->setEnabled(true)
                ->setOptions(['fields' => []])
                ->setFormOptions(['type' => 'contains']) // type of string filtering option, if you one to have just one
        )
        ->addActionGroup(MainActionGroup::create(
            Action::create('create', 'create')
                ->setLabel('sylius.ui.create')
                ->setEnabled(true)
                ->setIcon('plus')
                ->setPosition(100)
                ->setOptions([]),
        ))
        ->addActionGroup(
            ItemActionGroup::create(
                Action::create('update', 'update')
                    ->setLabel('sylius.ui.edit')
                    ->setEnabled(true)
                    ->setIcon('pencil')
                    ->setPosition(100)
                    ->setOptions([]),
                Action::create('delete', 'delete')
                    ->setLabel('sylius.ui.delete')
                    ->setEnabled(true)
                    ->setIcon('trash')
                    ->setPosition(100)
                    ->setOptions([]),
                Action::create('show', 'show')
                    ->setLabel('sylius.ui.show')
                    ->setEnabled(true)
                    ->setIcon('search')
                    ->setPosition(100)
                    ->setOptions([
                        'link' => [
                            'route' => 'app_user_show',
                            'parameters' => [
                                'id' => 'resource.id',
                            ],          
                        ],          
                    ]),
                Action::create('archive', 'archive')
                    ->setLabel('sylius.ui.archive')
                    ->setEnabled(true)
                    ->setIcon('search')
                    ->setPosition(100)
                    ->setOptions([
                        'restore_label' => 'sylius.ui.restore',          
                    ]),
            )
        )
        ->addActionGroup(
            BulkActionGroup::create(
                Action::create('delete', 'delete')
                    ->setLabel('sylius.ui.delete')
                    ->setEnabled(true)
                    ->setIcon('trash')
                    ->setPosition(100)
                    ->setOptions([]),
            )
        )
        ->addActionGroup(
            SubItemActionGroup::create(
                Action::create('addresses', 'links')
                    ->setLabel('sylius.ui.manage_addresses')
                    ->setOptions([
                        'icon' => 'cubes',
                        'links' => [
                            'index' => [
                                'label' => 'sylius.ui.list_addresses',
                                'icon' => 'list',
                                'route' => 'app_admin_user_address_index',
                                'visible' => 'resource.hasAddress',
                                'parameters' => [
                                    'userId' => 'resource.id',
                                ],
                            ],
                            'create' => [
                                'label' => 'sylius.ui.generate',
                                'icon' => 'random',
                                'route' => 'app_admin_user_address_create',
                                'parameters' => [
                                    'userId' => 'resource.id',
                                ],
                            ],
                        ],
                    ]),
            )
        )
    );
};
```

</details>


CHANGELOG.md
# CHANGELOG

## v1.13.1 (2024-11-29)

#### Details

- [#347](https://github.com/Sylius/SyliusGridBundle/issues/347) [Fix] Flip ([@loic425](https://github.com/loic425))
- [#353](https://github.com/Sylius/SyliusGridBundle/issues/353) Update interface name in custom_filter docs ([@arti0090](https://github.com/arti0090))
- [#350](https://github.com/Sylius/SyliusGridBundle/issues/350) Update LICENSE year ([@damonsson](https://github.com/damonsson))
- [#351](https://github.com/Sylius/SyliusGridBundle/issues/351) Create LICENSE_OF_TRADEMARK_AND_LOGO ([@damonsson](https://github.com/damonsson))
- [#345](https://github.com/Sylius/SyliusGridBundle/issues/345) Update RegisterFiltersPass.php ([@dotdevio](https://github.com/dotdevio))
- [#355](https://github.com/Sylius/SyliusGridBundle/issues/355) Fix static analysis in RegisterFiltersPass ([@GSadee](https://github.com/GSadee))

## v1.13.0 (2024-10-08)

#### Details

- [#344](https://github.com/Sylius/SyliusGridBundle/issues/344) [Composer] Bump sylius/resource-bundle dev dependency to last stable release ([@GSadee](https://github.com/GSadee))

## v1.13.0-BETA.1 (2024-10-01)

#### Details

- [#343](https://github.com/Sylius/SyliusGridBundle/issues/343) [Composer] Bump ResourceBundle to 1.12 beta ([@GSadee](https://github.com/GSadee))
- [#342](https://github.com/Sylius/SyliusGridBundle/issues/342) Add storage for filters ([@loic425](https://github.com/loic425), [@GSadee](https://github.com/GSadee))

## v1.13.0-ALPHA.3 (2024-09-24)

#### Details

- [#320](https://github.com/Sylius/SyliusGridBundle/issues/320) [Docs] Fix typo suplier => supplier ([@nclsHart](https://github.com/nclsHart))
- [#341](https://github.com/Sylius/SyliusGridBundle/issues/341) [Composer] Bump sylius/resource-bundle to 1.12 alpha version ([@GSadee](https://github.com/GSadee))
- [#319](https://github.com/Sylius/SyliusGridBundle/issues/319) [Gitattributes] Add tests directory to ignore list ([@loic425](https://github.com/loic425))
- [#302](https://github.com/Sylius/SyliusGridBundle/issues/302) Allow Pagerfanta 4.0 ([@mbabker](https://github.com/mbabker))

## v1.13.0-ALPHA.2 (2024-09-05)

#### Details

- [#146](https://github.com/Sylius/SyliusGridBundle/issues/146) Added NumericRange filter which gives the possibility of filtering … ([@christopherhero](https://github.com/christopherhero))
- [#328](https://github.com/Sylius/SyliusGridBundle/issues/328) Bump resource bundle version on dev requirements ([@loic425](https://github.com/loic425))
- [#327](https://github.com/Sylius/SyliusGridBundle/issues/327) Test grids with UI instead of API ([@loic425](https://github.com/loic425))
- [#323](https://github.com/Sylius/SyliusGridBundle/issues/323) Add timezone option to DatetimeFieldType ([@coldic3](https://github.com/coldic3))
- [#324](https://github.com/Sylius/SyliusGridBundle/issues/324) Add support for Symfony 7 ([@loic425](https://github.com/loic425))

## v1.13.0-ALPHA.1 (2024-01-30)

#### Details

- [#274](https://github.com/Sylius/SyliusGridBundle/issues/274) [GridBuilder] Add missing default value on filter ([@loic425](https://github.com/loic425))
- [#278](https://github.com/Sylius/SyliusGridBundle/issues/278) Add QueryBuilder access to custom filters ([@loic425](https://github.com/loic425))
- [#282](https://github.com/Sylius/SyliusGridBundle/issues/282) Add memberOf() method ([@TheMilek](https://github.com/TheMilek))
- [#279](https://github.com/Sylius/SyliusGridBundle/issues/279) Add services aliases to improve autowiring ([@loic425](https://github.com/loic425), [@lchrusciel](https://github.com/lchrusciel))
- [#276](https://github.com/Sylius/SyliusGridBundle/issues/276) Autoconfigure grid filters ([@loic425](https://github.com/loic425))
- [#287](https://github.com/Sylius/SyliusGridBundle/issues/287) Add missing filter to docs ([@TheMilek](https://github.com/TheMilek))
- [#291](https://github.com/Sylius/SyliusGridBundle/issues/291) Fix memberOf definition ([@TheMilek](https://github.com/TheMilek))
- [#301](https://github.com/Sylius/SyliusGridBundle/issues/301) Get Grid by FQCN ([@loic425](https://github.com/loic425))
- [#305](https://github.com/Sylius/SyliusGridBundle/issues/305) [GridBuilder] Apply transition action ([@loic425](https://github.com/loic425))
- [#299](https://github.com/Sylius/SyliusGridBundle/issues/299) Custom data provider (needs new resource metadata to detach the resource from Doctrine) ([@loic425](https://github.com/loic425))
- [#306](https://github.com/Sylius/SyliusGridBundle/issues/306) Fix build ([@loic425](https://github.com/loic425))
- [#307](https://github.com/Sylius/SyliusGridBundle/issues/307) Move tests app ([@loic425](https://github.com/loic425))
- [#308](https://github.com/Sylius/SyliusGridBundle/issues/308) [Maintenance] Launch of `Sylius Sp. z o.o.` copyright owner across project files ([@Rafikooo](https://github.com/Rafikooo))

## v1.12.2 (2024-10-08)

#### Details

- [#320](https://github.com/Sylius/SyliusGridBundle/issues/320) [Docs] Fix typo suplier => supplier ([@nclsHart](https://github.com/nclsHart))

## v1.12.1 (2024-01-05)

#### Details

- [#277](https://github.com/Sylius/SyliusGridBundle/issues/277) Fix the build ([@loic425](https://github.com/loic425))
- [#281](https://github.com/Sylius/SyliusGridBundle/issues/281) Init .gitattributes to preserve the planet ([@loic425](https://github.com/loic425))
- [#280](https://github.com/Sylius/SyliusGridBundle/issues/280) [docs] Add missing fields options on string filter ([@loic425](https://github.com/loic425))
- [#298](https://github.com/Sylius/SyliusGridBundle/issues/298) Fix CI ([@loic425](https://github.com/loic425))
- [#317](https://github.com/Sylius/SyliusGridBundle/issues/317) Fix coding standard ([@loic425](https://github.com/loic425))
- [#315](https://github.com/Sylius/SyliusGridBundle/issues/315) [Docs] Add info about available form options ([@loic425](https://github.com/loic425))

src/Bundle/DependencyInjection/Configuration.php
<?php

/*
 * This file is part of the Sylius package.
 *
 * (c) Sylius Sp. z o.o.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

declare(strict_types=1);

namespace Sylius\Bundle\GridBundle\DependencyInjection;

use Sylius\Bundle\GridBundle\Doctrine\ORM\Driver as DoctrineORMDriver;
use Sylius\Bundle\GridBundle\SyliusGridBundle;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

final class Configuration implements ConfigurationInterface
{
    public function getConfigTreeBuilder(): TreeBuilder
    {
        $treeBuilder = new TreeBuilder('sylius_grid');

        /** @var ArrayNodeDefinition $rootNode */
        $rootNode = $treeBuilder->getRootNode();

        $this->addDriversSection($rootNode);
        $this->addTemplatesSection($rootNode);
        $this->addGridsSection($rootNode);

        return $treeBuilder;
    }

    private function addDriversSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('drivers')
                    ->defaultValue([])
                    ->enumPrototype()->values(SyliusGridBundle::getAvailableDrivers())->end()
                ->end()
            ->end()
        ;
    }

    private function addTemplatesSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('templates')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->arrayNode('filter')
                            ->useAttributeAsKey('name')
                            ->scalarPrototype()->end()
                        ->end()
                        ->arrayNode('action')
                            ->useAttributeAsKey('name')
                            ->scalarPrototype()->end()
                        ->end()
                        ->arrayNode('bulk_action')
                            ->useAttributeAsKey('name')
                            ->scalarPrototype()->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;
    }

    private function addGridsSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('grids')
                    ->useAttributeAsKey('code')
                    ->arrayPrototype()
                        ->children()
                            ->scalarNode('extends')->cannotBeEmpty()->end()
                            ->arrayNode('driver')
                                ->addDefaultsIfNotSet()
                                ->children()
                                    ->scalarNode('name')->cannotBeEmpty()->defaultValue(DoctrineORMDriver::NAME)->end()
                                    ->arrayNode('options')
                                        ->performNoDeepMerging()
                                        ->variablePrototype()->end()
                                        ->defaultValue([])
                                    ->end()
                                ->end()
                            ->end()
                            ->arrayNode('sorting')
                                ->performNoDeepMerging()
                                ->useAttributeAsKey('name')
                                ->enumPrototype()->values(['asc', 'desc'])->cannotBeEmpty()->end()
                            ->end()
                            ->arrayNode('limits')
                                ->performNoDeepMerging()
                                ->integerPrototype()->end()
                                ->defaultValue([10, 25, 50])
                            ->end()
                            ->arrayNode('fields')
                                ->useAttributeAsKey('name')
                                ->arrayPrototype()
                                    ->children()
                                        ->scalarNode('type')->isRequired()->cannotBeEmpty()->end()
                                        ->scalarNode('label')->cannotBeEmpty()->end()
                                        ->scalarNode('path')->cannotBeEmpty()->end()
                                        ->scalarNode('sortable')->end()
                                        ->scalarNode('enabled')->defaultTrue()->end()
                                        ->scalarNode('position')->defaultValue(100)->end()
                                        ->arrayNode('options')
                                            ->performNoDeepMerging()
                                            ->variablePrototype()->end()
                                        ->end()
                                    ->end()
                                ->end()
                            ->end()
                            ->arrayNode('filters')
                                ->useAttributeAsKey('name')
                                ->arrayPrototype()
                                    ->children()
                                        ->scalarNode('type')->isRequired()->cannotBeEmpty()->end()
                                        ->scalarNode('label')->cannotBeEmpty()->end()
                                        ->scalarNode('enabled')->defaultTrue()->end()
                                        ->scalarNode('template')->end()
                                        ->scalarNode('position')->defaultValue(100)->end()
                                        ->arrayNode('options')
                                            ->performNoDeepMerging()
                                            ->variablePrototype()->end()
                                        ->end()
                                        ->arrayNode('form_options')
                                            ->performNoDeepMerging()
                                            ->variablePrototype()->end()
                                        ->end()
                                        ->variableNode('default_value')->end()
                                    ->end()
                                ->end()
                            ->end()
                            ->arrayNode('actions')
                                ->useAttributeAsKey('name')
                                ->arrayPrototype()
                                    ->useAttributeAsKey('name')
                                    ->arrayPrototype()
                                        ->children()
                                            ->scalarNode('type')->isRequired()->end()
                                            ->scalarNode('label')->end()
                                            ->scalarNode('enabled')->defaultTrue()->end()
                                            ->scalarNode('icon')->end()
                                            ->scalarNode('position')->defaultValue(100)->end()
                                            ->arrayNode('options')
                                                ->performNoDeepMerging()
                                                ->variablePrototype()->end()
                                            ->end()
                                        ->end()
                                    ->end()
                                ->end()
                            ->end()
                            ->arrayNode('removals')
                                ->ignoreExtraKeys(false)
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;
    }
}
